src/test/net_tests.cpp

```diff
@@ -769,4 +846,62 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
     g_mock_deterministic_tests = false;
 }
 
+BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
+{
+    // Pretend that we bound to this port.
+    const uint16_t bind_port = 20001;
+    gArgs.ForceSetArg("-bind", strprintf("3.4.5.6:%u", bind_port));
+
+    // Our address:port as seen from the peer.
+    in_addr peer_us_addr;
+    peer_us_addr.s_addr = htonl(0x02030405);
+    const CAddress peer_us{CService{peer_us_addr, 20002}, NODE_NETWORK};
+
+    // Create a peer with a routable IPv4 address.
+    in_addr peer_in_addr;
+    peer_in_addr.s_addr = htonl(0x01020304);
+    CNode peer{0,
+               NODE_NETWORK,
+               0,
+               INVALID_SOCKET,
+               CAddress{CService{peer_in_addr, 8333}, NODE_NETWORK},
+               0,
+               0,
+               CAddress{},
+               std::string{},
+               ConnectionType::OUTBOUND_FULL_RELAY};
+
+    const uint64_t services{NODE_NETWORK | NODE_WITNESS};
+    const int64_t time{0};
+    // Our address, as seen from the peer.
+    const CNetMsgMaker msg_maker{PROTOCOL_VERSION};
+    const auto msg = msg_maker.Make(NetMsgType::VERSION, PROTOCOL_VERSION, services, time, peer_us);
+
+    CDataStream recv_stream{msg.data, SER_NETWORK, PROTOCOL_VERSION};
+
+    // Force CChainState::IsInitialBlockDownload() to return false.
+    const auto nMinimumChainWork_orig = nMinimumChainWork;
+    const auto nMaxTipAge_orig = nMaxTipAge;
+    nMinimumChainWork = 0;
+    nMaxTipAge = GetTime();
```

Comment:
Can be written shorter

```suggestion
    generatetoaddress(m_node, ADDRESS_BCRT1_UNSPENDABLE);
```

Reply:
`CChainState::IsInitialBlockDownload()` still returns `true` :/

(it was necessary to `s/TestingSetup/RegTestingSetup/` in `BOOST_FIXTURE_TEST_SUITE()` in order to get `generatetoaddress()` to accept `ADDRESS_BCRT1_UNSPENDABLE`)

Reply:
Oh, sorry my bad. The mined block will have the timestamp of genesis (plus one), so the time is still too far back.

An alternative would be to use mocktime, also ugly, but less LOC.

```diff
diff --git a/src/test/net_tests.cpp b/src/test/net_tests.cpp
index 9519f3fd0c..7e77b758b2 100644
--- a/src/test/net_tests.cpp
+++ b/src/test/net_tests.cpp
@@ -88,7 +88,7 @@ static CDataStream AddrmanToStream(CAddrManSerializationMock& _addrman)
     return CDataStream(vchData, SER_DISK, CLIENT_VERSION);
 }
 
-BOOST_FIXTURE_TEST_SUITE(net_tests, TestingSetup)
+BOOST_FIXTURE_TEST_SUITE(net_tests, RegTestingSetup)
 
 BOOST_AUTO_TEST_CASE(cnode_listen_port)
 {
@@ -879,11 +879,10 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
 
     CDataStream recv_stream{msg.data, SER_NETWORK, PROTOCOL_VERSION};
 
-    // Force CChainState::IsInitialBlockDownload() to return false.
-    const auto nMinimumChainWork_orig = nMinimumChainWork;
-    const auto nMaxTipAge_orig = nMaxTipAge;
-    nMinimumChainWork = 0;
-    nMaxTipAge = GetTime();
+    // Force CChainState::IsInitialBlockDownload() to return false, otherwise PushAddress isn't called
+    SetMockTime(1);
+    BOOST_REQUIRE(!m_node.chainman->ActiveChainstate().IsInitialBlockDownload());
+    SetMockTime(0);
 
     m_node.peerman->InitializeNode(&peer);
 
@@ -898,9 +897,6 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
     BOOST_CHECK(std::any_of(peer.vAddrToSend.begin(), peer.vAddrToSend.end(),
                             [&expected](const CAddress& cur) { return cur == expected; }));
 
-    nMinimumChainWork = nMinimumChainWork_orig;
-    nMaxTipAge = nMaxTipAge_orig;
-
     gArgs.ForceSetArg("-bind", "");
 }

Reply:
Changed, less LOC. Thanks!